### DISCUSSION TITLE
Doesn't work on Debian 10

### DISCUSSION THREAD

**Discussion by thespicer21** [savonet/liquidsoap#1838] [Installation]
Hey guys, have installed this on Deb 9 with no problems in the past but cannot get it to work on Deb 10.8 at all.

`apt-get install liquidsoap` installs an old version 1.3.3, but there is no liquidsoap folder created at `/etc/liquidsoap`. Also there are no services associated to it when running `service liquidsoap status`. The program is installed somewhere as the help file opens up but I can't locate it anywhere?

Also looked for `/usr/local/lib/liquidsoap/` as per the docs but again, doesn't exist.

Tried installing as a sudo user first, then wiped it and tried as root. Neither worked.

So I removed all of that and tried downoading and installing the 1.4.4 deb file directly. This fails with a list of dependencies that no longer exist in the debian repo.

Also removed all of that and tried installing with opam but that is just a nightmare, didn't work at all, threw errors all over the place even after setting the correct version etc, didn't even bother noting them down as I never intend to use opam to install anything again - what a mess that is.

Does anybody have any other suggestions to get this working on Deb 10 please?

**Comment by oranjebloom** [comment]
Hi, I have liquidsoap version 1.4.4 running as a service on Debian 10. I installed it via opam by doing the following...

#install via opam#
```
opam list
opam update
opam upgrade 
opam switch create 4.08.0
```

```
opam init
eval `opam config env`
opam install depext
opam depext taglib mad lame cry ssl samplerate magic lastfm inotify liquidsoap
opam install taglib mad lame cry ssl samplerate magic lastfm inotify liquidsoap
```

You may need to add or remove some of the above (depending on what you need)

You then need to install the liquidsoap daemon ... I did it by doing the following...

```
git clone https://github.com/savonet/liquidsoap-daemon
cd liquidsoap-daemon
./daemonize-liquidsoap.sh PATH/TO/SCRIPT.liq
```

**Comment by thespicer21** [comment]
Thanks so much for your helpful reply oranjebloom!

I've managed to get it installed and running now but I still can't manage to get it running as a service.

```
debian@ns3070206:~$ service status liquidsoap
status: unrecognized service

debian@ns3070206:~$ sudo service liquidsoap start
Failed to start liquidsoap.service: Unit liquidsoap.service not found.
```

Any ideas how I would achieve this?

**Comment by oranjebloom** [comment]
I run my service by doing `sudo systemctl start NAME.liq-liquidsoap` where NAME is the name of your liq script.
Try that first.

If that doesn't work we then need to look at where your `liq` files are located on your system.
In my case, I created a folder named _script_ and put my `.liq`files in there ....so mine are located in _liquidsoap-daemon/scripts_

Then from the _liquidsoap-daemon/_ directory, I ran the daemonize command 
`./daemonize-liquidsoap.sh NAME.liq`  (again, NAME is the name of your liq script. You don't need to add a path, just the name of the file)

Doing that should create a new file ending in `.liq-run.liq` in the _liquidsoap-daemon/script_ directory. 
It will also create files ending in `.liq-liquidsoap.logrotate` and `.liq-liquidsoap.systemd` in the _liquidsoap-daemon_ directory. 

If you can see all these files in these locations after doing the steps above, then try running the service command again.
`sudo systemctl start NAME.liq-liquidsoap`

If this doesn't work, delete all those files, and try it a second time. It took me a few goes to get it working.

**Comment by oranjebloom** [comment]
Might be a problem with your script. Did the script work when you ran it in a previous version? Was it version 1.3.3? There may have been changes that cause it to break in version 1.4.4. Can you show me your script?

**Comment by thespicer21** [comment]
Thanks once again for your help, really appreciate it.

That is indeed the way to get the service to work, but sadly it's throwing up a whole bunch of other errors and I don't even know where to start...

Firstly when I try to start it I get this..

```
Job for radio.liq-liquidsoap.service failed because the control process exited with error code.
See "systemctl status radio.liq-liquidsoap.service" and "journalctl -xe" for details.
```

So, I do as it asks and get this...

```
debian@ns3070206:~$ sudo systemctl status radio.liq-liquidsoap.service
● radio.liq-liquidsoap.service - radio.liq Liquidsoap daemon
   Loaded: loaded (/etc/systemd/system/radio.liq-liquidsoap.service; disabled; vendor preset: enabled)
   Active: failed (Result: exit-code) since Mon 2021-03-29 02:49:17 UTC; 42s ago
     Docs: http://liquidsoap.fm/
  Process: 17976 ExecStart=/home/debian/.opam/4.08.0/bin/liquidsoap /home/debian/liquidsoap-daemon/script/radio.liq-run.liq (code=exited, status=1/FAILURE)

Mar 29 02:49:17 ns3070206 systemd[1]: radio.liq-liquidsoap.service: Service RestartSec=100ms expired, scheduling restart.
Mar 29 02:49:17 ns3070206 systemd[1]: radio.liq-liquidsoap.service: Scheduled restart job, restart counter is at 5.
Mar 29 02:49:17 ns3070206 systemd[1]: Stopped radio.liq Liquidsoap daemon.
Mar 29 02:49:17 ns3070206 systemd[1]: radio.liq-liquidsoap.service: Start request repeated too quickly.
Mar 29 02:49:17 ns3070206 systemd[1]: radio.liq-liquidsoap.service: Failed with result 'exit-code'.
Mar 29 02:49:17 ns3070206 systemd[1]: Failed to start radio.liq Liquidsoap daemon.
```
Then I try to run it directly in daemon mode and discover that I'm missing directories..

```
debian@ns3070206:~/liquidsoap-daemon$ liquidsoap radio.liq --daemon
FATAL ERROR: PID directory "/home/debian/.opam/4.08.0/lib/liquidsoap/var/run/liquidsoap" does not exist.
To change it, add the following to your script:
  set("init.daemon.pidfile.path", "<path>")
```

All I have in `/home/debian/.opam/4.08.0/lib/liquidsoap/` is : 

```
etc
share
```

No `var` dir or it's contents?

**Comment by oranjebloom** [comment]
Do you have this at the top of your liq script?

`#!/home/db/.opam/1.4.4/bin/liquidsoap`

**Comment by oranjebloom** [comment]
The path seems wrong. It should point to `bin` not `lib`

**Comment by oranjebloom** [comment]
can you also open up the file ending in `.liq-liquidsoap.systemd`  in the `liquidsoap-daemon` directory. What does that show?

**Comment by thespicer21** [comment]
Top of liq was :

`#!/usr/bin/liquidsoap`  doh!

Contents of `.liq-liquidsoap.systemd` :


```
[Unit]
Description=radio.liq Liquidsoap daemon
After=network.target
Documentation=http://liquidsoap.fm/

[Service]
Type=forking
User=debian
PIDFile=/home/debian/liquidsoap-daemon/pid/radio.liq-run.pid
WorkingDirectory=/home/debian/liquidsoap-daemon
ExecStart=/home/debian/.opam/4.08.0/bin/liquidsoap /home/debian/liquidsoap-daemon/script/radio.liq-run.liq
Restart=always

[Install]
WantedBy=multi-user.target
Alias=radio.liq-liquidsoap.service
```

**Comment by oranjebloom** [comment]
I recommend that you change your liq script so this is at the top ...

#!/home/debian/.opam/4.08.0/bin/liquidsoap/

...then delete the files ending in ...
```
.liq-run.liq 
.liq-liquidsoap.systemd
.liq-liquidsoap.logrotate
```

Then daemonize your updated liq script by running `./daemonize-liquidsoap.sh NAME.liq` again from the _liquidsoap-damon_ directory

after that try to start the daemon

**Comment by thespicer21** [comment]
This is my .liq now : 

```
#!/home/debian/.opam/4.08.0/bin/liquidsoap/
# Log dir
set("log.file.path","/tmp/basic-radio.log")

# Define The Mountpoints
ices = mksafe(input.http("http://localhost:8000/ices"))
live = input.http("http://localhost:8000/live")

# Playout Priority
radio = fallback(track_sensitive=false,
          [live,ices])

output.icecast(%mp3(bitrate=192),host="localhost",port=8000,
password="mp31922019",
description="test",
url="https://test.com",
mount="liquidsoap",name="test",radio)
```

Deleted `.liq-liquidsoap.systemd `& `.liq-liquidsoap.logrotate` but didn't have a `.liq-run.liq ` in there.

Re-deamonized it...

Tried to start it again and got the same error message, when I do status again I get the exact same report as above?

**Comment by oranjebloom** [comment]
where exactly is your liq script located? Is it in /home/debian/liquidsoap-daemon/script/?

**Comment by thespicer21** [comment]
It's at `/home/debian/liquidsoap-daemon/radio.liq`

**Comment by thespicer21** [comment]
This is everything in there : 

```
-rw-r--r-- 1 debian debian    504 Mar 27 18:51 CHANGES
-rwxr-xr-x 1 debian debian   4036 Mar 27 18:51 daemonize-liquidsoap.sh
-rw-r--r-- 1 debian debian   1409 Mar 27 18:51 liquidsoap.initd.in
-rw-r--r-- 1 debian debian    541 Mar 27 18:51 liquidsoap.launchd.in
-rw-r--r-- 1 debian debian    339 Mar 27 18:51 liquidsoap.logrotate.in
-rw-r--r-- 1 debian debian    326 Mar 27 18:51 liquidsoap.systemd.in
drwxr-xr-x 2 debian debian   4096 Mar 27 18:59 log
-rw-r--r-- 1 debian debian   1187 Mar 27 18:51 opam
drwxr-xr-x 2 debian debian   4096 Mar 27 18:59 pid
-rw-r--r-- 1 debian debian    526 Mar 29 03:17 radio.liq
-rw-r--r-- 1 debian debian    399 Mar 29 03:19 radio.liq-liquidsoap.logrotate
-rw-r--r-- 1 debian debian    444 Mar 29 03:19 radio.liq-liquidsoap.systemd
-rw-r--r-- 1 debian debian   2056 Mar 27 18:51 README.md
drwxr-xr-x 2 debian debian   4096 Mar 27 19:05 script
```

**Comment by thespicer21** [comment]
I'll move it into script and re-daemozine again

**Comment by oranjebloom** [comment]
create a script directory inside liquidsoap-daemon and place the liq script inside there. Then delete all the files like before and daamonize the script again

**Comment by oranjebloom** [comment]
yep

**Comment by thespicer21** [comment]
Ahh there is the other .liq file. :D

**Comment by oranjebloom** [comment]
after you have daemonised the liq file in the `script` directory  you should see the `.liq-run.liq ` next to the liq file in that script directory. The .`liq-liquidsoap.systemd` and .`liq-liquidsoap.logrotate` will be a level up in the _liquidsoap-daemon_ diretcory

**Comment by thespicer21** [comment]
Finally got there in the end, I tried doing `./daemonize-liquidsoap.sh script/radio.liq` but that obvioulsy didn't work as I'm sure you know. I wasn't aware that the .sh is looking in that script folder for the goodness already ;)

So `./daemonize-liquidsoap.sh radio.liq` did the job...

You sir are a legend. Thank you so, so much. It's all working now. I owe you a few beers ;)

**Comment by thespicer21** [comment]
Oh one final thing before I leave you in peace... Is this "not our child" msg normal?

`Mar 29 03:41:23 ns3070206 systemd[1]: Starting radio.liq Liquidsoap daemon...
Mar 29 03:41:23 ns3070206 systemd[1]: radio.liq-liquidsoap.service: Supervising process 19643 which is not our child. We'll most likely not notice when it exits.
Mar 29 03:41:23 ns3070206 systemd[1]: Started radio.liq Liquidsoap daemon.`

**Comment by oranjebloom** [comment]
No worries. Glad you got there. It takes some effort to set up but satisfying once you get it all going. I've had those "not our child" message too but I think they are safe to ignore. It's something that perhaps @toots could answer.

**Comment by toots** [comment]
Hey y'all. Thanks for helping @oranjebloom ! 

I wonder: do y'all use the `set("init.daemon", true)` settings? I believe that `systemd` can handles this side of the things so it shouldn't be used in your liquidsoap script. If used, the process would be forked away, which could cause the kind of issue that is described in the `systemd` logs.

**Comment by oranjebloom** [comment]
Hi @toots ... no I don't include that setting in my script. So then, you recommend simply adding `set("init.daemon", true)` to the liq script and then daemonize that script?  Are any other changes necessary? If there is any documentation about it, please let me know.

**Comment by oranjebloom** [comment]
Oh, ignore the previous reply. I misread your comment. In short, no I don't use`set("init.daemon", true)` in my script

**Comment by toots** [comment]
Well actually this is a good question. I thought `liquidsoap-daemon` was launching liquidsoap not as a daemon but it looks like it's doing. That could be the cause of your issue.

Would you mind checking if changing that helps it?  There should be a `<radio_name>-run.liq` script with all the settings used to launch the daemon. If you have this:
```
set("init.daemon",true)
set("init.daemon.change_user",true)
set("init.daemon.change_user.group","${USER}")
set("init.daemon.change_user.user","${USER}")
set("init.daemon.pidfile",true)
set("init.daemon.pidfile.path","${pid_dir}/${script_name}-run.pid")
```
Could you try and comment it out and see if that helps? Thanks!

**Comment by hayden-t** [comment]
I was able to get my/the-old /etc/init.d/liquidsoap startup daemon script from the debian stretch package and it still works even though im running buster now
